Add basic instructions on how to use Split for feature flags (#787)
mattermost/mattermost-developer-documentation: site/content/contribute/server/feature-flags.md

- Tests should be written to verify the feature flag works as expected. Note that in cases where there may be a migration or new data, off to on and on to off should both be tested.
- Log messages by the feature should include the feature flag tag, with the feature flag name as a value, in order to ease debugging.

## Split

> [Split](https://split.io) is the feature flag management service we use for internal testing and our production Cloud service. It enables us to create and control feature flags for multiple servers and environments as well as the ability to run A/B split tests.

Reviewer: ```suggestion
[Split](https://split.io) is the feature flag management service we use for internal testing and for our production Cloud service. It enables us to create and control feature flags for multiple servers and environments as well as gives us the ability to run A/B split tests.
```

> We've used their SDK to integrate with the Mattermost server. You can find the [Split documentation here](https://help.split.io/hc/en-us).

Reviewer: ```suggestion
We've used Split's SDK to integrate with the Mattermost server. You can find the [Split documentation here](https://help.split.io/hc/en-us).
```

### Who has access to Split?

> We currently have a limited number of seats for our Split account so not everyone gets access. Currently, 2 people per engineering team plus some PM/UX folk will have access. If you need to manage some feature flags, ask your team lead who has access and can help you out. It's up to each team to determine who has access.

Reviewer: ```suggestion
We currently have a limited number of seats for our Split account, so not everyone gets access. Currently, two people per each engineering team plus some PM/UX folk have access. If you need to manage some feature flags, ask your team lead who has access and can help you out. It's up to each team to determine who has access.
```

> ### How to log in?

Reviewer: ```suggestion
### How do I log in?
```

Login is controlled via OneLogin. If you have access you will be able to log in through there.

> ### Terminology

Reviewer 2: ```suggestion

### Terminology
```

In Split, there's some terminology you will need to understand:

 - **Split** - A feature flag
 - **Treatment** - The values a split (e.g. feature flag) can have

>  - **Environment** - Different environments that splits can target such as test, staging and production

Reviewer: ```suggestion
 - **Environment** - Different environments that splits can target such as test, staging, and production
```

 - **Server** - Single Mattermost workspace
 - **Segment** - A list of servers that can be targeted by a split

### Adding a split

After you've added the feature flag in code, you'll need to create a split for it.

 1. Log in to Split.

>  2. On left sidebar, click "Splits".

Reviewer: ```suggestion
 2. On left sidebar, click **Splits**.
```

>  3. At the top click "Create split".

Reviewer: ```suggestion
 3. At the top click **Create split**.
```

 4. Fill in the following:

>    1. Enter the name of your feature flag. It must match the name used in code in `model/feature_flags.go`.
>    2. Select "server" as the traffic type.
>    3. Optionally add tags and a description.
>    4. Click "Create".

Reviewer: ```suggestion
     1. Enter the name of your feature flag. It must match the name used in code in `model/feature_flags.go`.
     2. Select ``server`` as the traffic type.
     3. Optionally add tags and a description.
     4. Click **Create**.
```

The split is now created but not targeting anything. The next section covers adding basic targeting rules.

> ###  Add targeting rules to a split

Reviewer 3: During the 101 training, the CSM mentioned people getting tripped up by the hierarchy of targeting rules. E.g. if a server ID is listed as a targeted server in rule 2, they will still get the first treatment if they satisfy the requirements for rule 1.

Not sure if we want to / need to explicitly mention it here.

Author: Hmm I'd prefer to keep our docs simple so that there's less to maintain but what about linking out to the split docs for more info for things like this?

Reviewer 3: I don't feel strongly about it, but it was the one common pitfall the CSM mentioned people experience.

To have the split have any effect, it needs targeting rules to know which environment and servers it should set the feature flag value to. To add basic targeting rules:

 1. Find your split and click on it.

>  2. At the top, under the name of the split select the "Environment" you want to target.

Reviewer: ```suggestion
 2. Under the name of the split at the top, select the **Environment** that you want to target.
```

>    - See [Environments](#Environments) for a description of each.
>  3. In the center of the screen, select "Add Rules".
>  4. Define your treatments:
>    - Recall that treatments are just the values your feature flag can have.
>    - Boolean flags should use the default "on" and "off" treatments.
>    - For non-boolean flags, change the treatments to be the possible values. E.g. for a split that specifies a plugin's version you could have three treatments of "1.0", "1.1" and "2.0".
>  5. Down the page a bit, set the default rule to one of your treatments.
>    - This will set the flag for all servers in the environment while you don't have any other targeting rules configured.
>  6. At the top right, click "Save changes" and confirm them on the next screen.

Reviewer: ```suggestion
     - See [Environments](#Environments) for a description of each.
 3. In the center of the screen, select **Add Rules**.
 4. Define your treatments:
     - Recall that treatments are just the values your feature flag can have.
     - Boolean flags should use the default "on" and "off" treatments.
     - For non-boolean flags, change the treatments to be the possible values. E.g. for a split that specifies a plugin's version you could have three treatments of "1.0", "1.1" and "2.0".
 5. Down the page a bit, set the default rule to one of your treatments.
     - This will set the flag for all servers in the environment while you don't have any other targeting rules configured.
 6. At the top right, click **Save changes** and confirm them on the next screen.
```

That covers targeting an environment with a split and adding a default rule for all servers in that environment. The next section covers targeting specific servers.

### Target a specific server

> The ID for a "server" as seen by Split is the Mattermost server's telemetry ID. To target a specific server you need that telemetry ID. The main way to get the telemetry ID from a server would be from the analytics but you can also manually grab the telemetry ID by one of the following methods:

Reviewer: "...would be from the analytics..." - Analytics team?

Author: Not exactly sure how or where it's exposed, maybe it's a new dashboard or search in Looker. @enelson720 or @jasonblais any idea for how we'd be able to give people the ability to find the telemetry ID from the analytics data?

Reviewer 3: Telemetry ID is the same as a server ID. @enelson720 will defer to you on the exact steps to retrieve from Looker.

Reviewer 4: Yeah, it's fairly straight forward @jwilander. I can provide a look that can be filtered by customer name and/or workspace name in the case of cloud workspaces. [Here's a jumping off point for retrieving the telemetry id.](https://mattermost.looker.com/explore/data_warehouse/server_fact?toggle=fil&qid=m250frAeAMARzhZBek5C4e)

> ##### Get telemetry ID via inspecting browser traffic

Reviewer 3: ```suggestion
#### Get telemetry ID via inspecting browser traffic
```

 1. In Chrome (any browser with dev tools will work but instructions are specific to Chrome), open the developer console and switch to the Network tab.
 2. Type "config" into the filter for the network requests.
 3. Go to the server's URL (e.g. xxxxxx.test.mattermost.cloud).
 4. In the network tab, click on the request for `/api/v4/config/client?format=old`.
 5. Click on the "Response" tab and search for "TelemetryId".
 6. Copy the value of "TelemetryId".

> ##### Get telemetry ID via curl and jq

Reviewer 3: ```suggestion
#### Get telemetry ID via curl and jq
```

```bash
$ curl -s https://<your-workspace-url>/api/v4/config/client?format=old | jq ".TelemetryId"
```

> #### Target using the telemetry ID

Reviewer 3: ```suggestion
### Target using the telemetry ID
```

Author: This is purposefully under the `Target a specific server` section

> With the telemetry ID, go to your split in Split and do the following:
> 
>  1. Find the section "Create whitelists" and click "Add whitelist".
>  2. Choose the treatment value you want the workspace to get.
>  3. Under "Server" paste in your telemetry ID.
>  4. At the top right, click "Save changes" and confirm them on the next screen.
> 
> Now your workspace will have it's feature flag set according to this rule, regardless of what the default for the split is set to.

Reviewer: ```suggestion
With the telemetry ID, go to your split and do the following:

 1. Find the section "Create whitelists" and click **Add whitelist**.
 2. Choose the treatment value you want the workspace to get.
 3. Under "Server" paste in your telemetry ID.
 4. At the top right, click **Save changes** and confirm them on the next screen.

Now your workspace will have its feature flag set according to this rule, regardless of what the default for the split is set to.
```

### Add or update a feature flag for community.mattermost.com

> Note that this will set the flag for community.mattermost.com, community-release.mattermost.com, and community-daily.mattermost.com. It is not possible to set a rule for only one.

Reviewer 5: I think there's a strong use case to allow to distinguish between those a bit more. I'd like to use community-daily for testing of more "unfinished but functional" features (feature flag on), and leave community stable with the feature flag off. Would it be possible to split these three out?

Not a comment specific to the doc, but to the setup.

Author: We can do it by setting env vars for the different deployments but it requires manual work from an SRE. To have split.io support it we'd need to do some hacky coding in the MM server since only the cluster leader syncs the flag right now and community, -release and -daily need to all be in the same cluster.

Author: Any idea of how else we might be able to do that @crspeller ?

Reviewer 6: No. Environment variables is the escape hatch for situations like community. If we want strange feature flag configurations that's how we do it.

 1. Open the split for your flag.
 2. At the top under the split name, select the "Cloud Staging" environment.

>  3. Find "Set targeting rules" and click **Add rule**.

Reviewer 6: Might want to comment that the targeting rule might already exist. Also isn't community the only thing on staging now? (i just set IC for the whole environment)

Author: Added a note about that. And yes right now the only thing hooked up to split in staging is community but in the future there could be other installations on staging using split too

>  1. Open the split for your flag.
>  2. At the top under the split name, select the "Cloud Staging" environment.
>  3. Find "Set targeting rules" and click "Add rule".
>  4. Leave the rule set to "is in segment" and click "Select Segment..." and choose "community".
>  5. Select the treatment you want community.mattermost.com to get.
>  6. At the top right, click "Save changes" and confirm them on the next screen.

Reviewer: ```suggestion
 1. Open the split for your flag.
 2. At the top under the split name, select the "Cloud Staging" environment.
 3. Find "Set targeting rules" and click **Add rule**.
 4. Leave the rule set to "is in segment" and click **Select Segment...** and choose "community".
 5. Select the treatment you want community.mattermost.com to get.
 6. At the top right, click **Save changes** and confirm them on the next screen.
```

The feature flag is now set for community.mattermost.com.

Syncing occurs once every 30 seconds so your feature flag will get changes at maximum 30 seconds after they're made.

### Environments

>  - Cloud Test - All the Cloud test servers/workspaces, including those made on PRs and by the /cloud command.
>  - Cloud Staging - Mainly community.mattermost.com, including community-release and community-daily. It's not possible to target those separately.
>  - Cloud Production - All our production Cloud workspaces.

Reviewer: ```suggestion
 - Cloud Test - All the Cloud test servers/workspaces, including those made on PRs and by the ``/cloud`` command.
 - Cloud Staging - Mainly community.mattermost.com, including community-release and community-daily. It's not possible to target those separately.
 - Cloud Production - All our production Cloud workspaces.
```

### Split FAQ

> >How long does it take for workspaces/servers to pick up on changes to feature flags in Split?

Reviewer: Incomplete or TBD?

Author: Hmm looks like my change didn't get committed. Fixing.

```
#### Target using the telemetry ID

With the telemetry ID, go to your split and do the following:

 1. Find the section "Create whitelists" and click **Add whitelist**.
 2. Choose the treatment value you want the workspace to get.
 3. Under "Server" paste in your telemetry ID.
 4. At the top right, click **Save changes** and confirm them on the next screen.

Now your workspace will have its feature flag set according to this rule, regardless of what the default for the split is set to.

### Add or update a feature flag for community.mattermost.com

Note that this will set the flag for community.mattermost.com, community-release.mattermost.com, and community-daily.mattermost.com. It is not possible to set a rule for only one via Split. It is possible to set the flags only for one version of community via environment variables. This will require manual configuration by the SRE team. You can ask for support for this on community.mattermost.com via the `@sresupport` mention.

To set a flag via Split for community:

 1. Open the split for your flag.
 2. At the top under the split name, select the "Cloud Staging" environment.
 3. Find "Set targeting rules" and click **Add rule**.

>    - The rule may already exist, in that case you can edit the existing one.
>  4. Leave the rule set to "is in segment" and click **Select Segment...** and choose "community".
>  5. Select the treatment you want community.mattermost.com to get.
>  6. At the top right, click **Save changes** and confirm them on the next screen.
> 
> The feature flag is now set for community.mattermost.com.

Reviewer: ```suggestion
   - The rule may already exist, and in that case you can edit the existing one.
 4. Leave the rule set to "is in segment" and click **Select Segment...** and choose "community".
 5. Select the treatment you want community.mattermost.com to get.
 6. At the top right, click **Save changes** and confirm them on the next screen.

The feature flag is now set for community.mattermost.com.
```

### Environments

>  - Cloud Test - All the Cloud test servers/workspaces, including those made on PRs and by the ``/cloud`` command.

Reviewer: ```suggestion
 - Cloud Test - All the Cloud test servers/workspaces, including those created on PRs and by the ``/cloud`` command.
```

 - Cloud Staging - Mainly community.mattermost.com, including community-release and community-daily. It's not possible to target those separately.
 - Cloud Production - All our production Cloud workspaces.

### Split FAQ

> >How long does it take for workspaces/servers to pick up on changes to feature flags in Split?
> 
> Syncing occurs once every 30 seconds so your feature flag will get changes at maximum 30 seconds after they're made.

Reviewer: ```suggestion
How long does it take for workspaces/servers to pick up on changes to feature flags in Split?
 - Syncing occurs once every 30 seconds so your feature flag will get changes at maximum 30 seconds after they're made.
```
